asm/sjis.s

```diff
@@ -0,0 +1,206 @@
+.include "macro.inc"
+
+# assembler directives
+.set noat      # allow manual use of $at
+.set noreorder # don't insert nops after branches
+.set gp=64     # allow use of 64-bit general purpose registers
+
+.section .text
+
+.balign 16
+
+# s32 Kanji_OffsetFromShiftJIS(s32 shiftJISCodepoint)
+# 
+# Returns the offset of the glyph texture data in `kanji` corresponding to a
+# given 2-byte Shift-JIS codepoint. No range validity check is carried out.
+#
+# A nice Shift-JIS codepoint table: https://uic.io/en/charset/show/shift_jis/
+# `kanji` contains the 'Level 1' kanji (889F-9872), and a reworked version of 
+# the non-kanji section.
+#
+# @param shiftJISCodepoint Codepoint of glyph.
+# return s32 offset into `kanji` file.
+BEGIN Kanji_OffsetFromShiftJIS
+    # Characters with codepoints >= 0x8800 are kanji. Arrangement is regular, 
+    # so convert index directly.
+    li      $at, 0x8800
+    slt     $at, $a0, $at
+    bnez    $at, .not_kanji
+     # 0xBC is the size of one block of glyphs in `kanji`:
+     # 0x100 possible codepoints with the same byte1
+     # - 0x40 unused at beginning 
+     # - 1 unused at 0x7F 
+     # - 3 unused at 0xFD, 0xFE, 0xFF
+     li     $a2, 0xBC
+    # Get byte1 and adjust so starts at 0
+    srl     $a1, $a0, 8
+    addi    $a1, $a1, -0x88
+    multu   $a2, $a1
+    # Get byte2 and adjust so starts at 0
+    andi    $a3, $a0, 0xff
+    addi    $a3, $a3, -0x40
+    slti    $at, $a3, 0x40
+    mflo    $a2
+    # 0xXX7F is always empty and elided in the file, so decrement
+    bnezl   $at, .kanji_lower_block
+     mflo   $a2
+    addi    $a3, $a3, -1
+    mflo    $a2
+.kanji_lower_block:
+    addi    $a3, $a3, 0x30a
+    add     $a3, $a3, $a2
+    jr      $ra
+     sll    $v0, $a3, 7
+     # returns (0x30A + (adjusted byte2) + (adjusted byte1) * 0xBC) * FONT_CHAR_TEX_SIZE
+
+# Non-kanji are arranged with irregular gaps, use the lookup table.
+.not_kanji:
+    # Get byte1 and adjust so starts at 0
+    srl     $a1, $a0, 8
+    addi    $a1, $a1, -0x81
+    multu   $a2, $a1
+    # Get byte2 and adjust so starts at 0
+    andi    $a3, $a0, 0xff
+    addi    $a3, $a3, -0x40
+    slti    $at, $a3, 0x40
+    mflo    $a2
+    # 0xXX7F is always empty and elided in the file, so decrement
+    bnezl   $at, .not_kanji_lower_block
+     mflo   $a2
+    addi    $a3, $a3, -1
+    mflo    $a2
+.not_kanji_lower_block:
+    add     $a3, $a3, $a2
+    lui     $a2, %hi(kanjiNonKanjiIndices)
+    sll     $a3, $a3, 1
+    addiu   $a2, %lo(kanjiNonKanjiIndices)
+    add     $a3, $a3, $a2
+    lh      $a2, ($a3)
+    jr      $ra
+     sll    $v0, $a2, 7
+     # returns kanjiNonKanjiIndices[(adjusted byte2) + (adjusted byte1) * 0xBC] * FONT_CHAR_TEX_SIZE
+END Kanji_OffsetFromShiftJIS
+
+# C pseudocode:
+# s32 Kanji_OffsetFromShiftJIS(s32 arg0) {
+#     s32 byte1;
+#     s32 byte2;
```

Comment:
```suggestion
#     s32 upperByte;
#     s32 lowerByte;
```
These names make more sense to me at least on how they graph the upper and lower bytes of `shiftJISCodepoint`

Reply:
I haven't found a particular pattern for how these are usually referred to for Japanese encodings: unlike Unicode, however you store it is always meant to be "big-endian", according to the standard: https://www.sljfaq.org/afaq/encodings.html , but I think as a communication standard they usually refer to first, second, third, etc. bytes (only the first 2 mattering), because while this function only uses 2-byte characters, technically Shift-JIS should cover the 1-byte ones as well (rather like how Unicode can use as many bytes as it wants for a character, but much simpler). Hence the way you identify characters is theoretically supposed to be from reading the stream of bytes rather than wider types (and yes, since this function doesn't do it, it is big-endian only without appropriate wrapping, although that's inevitable since it's handwritten MIPS).